bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/transactions/TransactionResourceManager.java

```diff
@@ -359,19 +486,38 @@ public TransactionLocalContext getCurrentTransactionContext() {
      */
     void endXATransaction(String transactionId, String transactionBlockId) {
         String combinedId = generateCombinedTransactionId(transactionId, transactionBlockId);
-        Transaction trx = trxRegistry.get(combinedId);
-        if (trx != null) {
-            List<BallerinaTransactionContext> txContextList = resourceRegistry.get(combinedId);
-            if (txContextList != null) {
-                for (BallerinaTransactionContext ctx : txContextList) {
-                    try {
-                        XAResource xaResource = ctx.getXAResource();
-                        if (xaResource != null) {
-                            trx.delistResource(xaResource, TMSUCCESS);
+        if (transactionManagerEnabled) {
+            Transaction trx = trxRegistry.get(combinedId);
+            if (trx != null) {
+                List<BallerinaTransactionContext> txContextList = resourceRegistry.get(combinedId);
+                if (txContextList != null) {
+                    for (BallerinaTransactionContext ctx : txContextList) {
+                        try {
+                            XAResource xaResource = ctx.getXAResource();
+                            if (xaResource != null) {
+                                trx.delistResource(xaResource, TMSUCCESS);
+                            }
+                        } catch (IllegalStateException | SystemException e) {
+                            log.error("error in ending the XA transaction " + transactionId
+                                    + ":" + e.getMessage(), e);
+                        }
+                    }
+                }
+            }
+        } else {
+            Xid xid = xidRegistry.get(combinedId);
+            if (xid != null) {
```

Comment:
merge xid and txContextList != null checks

Reply:
Fixed in 91ca624